internal/controller/api.go

```diff
@@ -107,20 +124,53 @@ func (c Controller) backoff() (time.Duration, time.Duration) {
 // Use the builder methods in this package (starting with ForType) to construct
 // a controller, and then pass it to a Manager to be executed.
 type Controller struct {
-	managedType *pbresource.Type
-	reconciler  Reconciler
-	logger      hclog.Logger
-	watches     []watch
-	baseBackoff time.Duration
-	maxBackoff  time.Duration
-	placement   Placement
+	managedType   *pbresource.Type
+	reconciler    Reconciler
+	logger        hclog.Logger
+	watches       []watch
+	customWatches []customWatch
+	baseBackoff   time.Duration
+	maxBackoff    time.Duration
+	placement     Placement
 }
 
 type watch struct {
 	watchedType *pbresource.Type
 	mapper      DependencyMapper
 }
 
+// Watch is responsible for watching for custom events from source and adding them to
+// the event queue.
+func (s *Source) Watch(ctx context.Context, add func(e Event)) error {
```

Comment:
This never returns an error so maybe we can omit the return type

Reply:
good catch

Reply:
turns out we need this to return an error so that it's type can be used with `group.Go()` which requires a func with a signature `f func() error`